master/buildbot/steps/source/git.py

```diff
@@ -62,6 +62,122 @@ def isTrueOrIsExactlyZero(v):
 ]
 
 
+class GitMirrorRepo(object):
+"""
+Example:
+    git_mirror = GitMirrorRepo("my-repo", workdir, {
+        'repourl': 'http://whereever/project.git',
+        'origin':  'origin',    # required
+        'alwaysUpdate': False,  # default True
+        'fetchTags': True,      # default False
+    })
+    
+    #...
+    steps.append(Git(mirrorRepo=git_mirror, ...))
+"""
+
+    def __init__(self, name, workdir=None, repo_config=None, lock=None):
+        if not repo_config:
+            bbconfig.error("GitMirrorRepo: must provide a 'repo_config' dict")
+
+        if not lock:
+            # Access to the mirror should be exclusive on each slave.
+            # Git probably could handle concurrent accesses, but we might
+            # clobber the mirror repo if we think it's corrupt. The only
+            # way to protect that kind of thing is to make this all exclusive.
+            # Besides, making this one step exclusive should not be a big deal.
+            lock = SlaveLock("%s-GitMirrorRepo" % name, maxCount=1)
+        
+        self.name = name
+        self.workdir = workdir
+        self.repo_config = repo_config
+        self.lock = lock
+
+    def getLocks(self):
+        return [self.lock]
+
+    @defer.inlineCallbacks
+    def _checkRemoteUrl(self, dovccmd, repo, git_config):
+        remote     = repo['origin']
+        remote_url = repo['repourl']
+
+        actual_url = git_config.get('remote.%s.url' % remote)
+
+        # Already good?
+        if actual_url==remote_url:
+            return
+
+        # If it's set, remove it; this will also remove any remote-tracking branches.
+        if actual_url:
+            command = ['remote', 'remove', remote]
+            yield dovccmd(command, abandonOnFailure=False)
+
+        # Add it back.
+        command = ['remote', 'add', remote, remote_url]
+        yield dovccmd(command, abandonOnFailure=False)
+
+    @defer.inlineCallbacks
+    def _checkRemoteTagOption(self, dovccmd, repo, git_config):
+        remote_tags = repo.get('fetchTags')
+
+        # Figure out what we expect it to be.
+        if remote_tags is None:
+            expect_tags = None
+        elif not remote_tags:
+            expect_tags = "--no-tags"
+        else:
+            expect_tags = "--tags"
+
+        remote_tags_key = 'remote.%s.url' % remote
+        if expect_tags==git_config.get(remote_tags_key):
+            return
+
+        if not expect_tags:
+            command = ['config', '--unset', remote_tags_key]
+            yield dovccmd(command, abandonOnFailure=False)
+        else:
+            command = ['config', remote_tags_key, expect_tags]
+            yield dovccmd(command, abandonOnFailure=False)
+
+    @defer.inlineCallbacks
+    def doCheckMirrorConfig(self, dovccmd):
+        """ Ensure the mirror's config is right. """
+
+        # Easiest to get the full config all at once.
+        git_config = yield dovccmd(['config', '--list', '-z'],
```

Comment:
No, it's actually fairly easy. 

In your master config, you would define one GitRepoMirror instance (to describe what the clone represents), and then pass that same instance as a param to each Git step (one or more builders). 

The idea is that the Mirror represents _one_ directory on a slave, and then each Git step clones that to its builder's workdir. 

The Mirror class has the job of maintaining that one directory (checking its config, fetching, etc), which is a bare clone since it is not directly used. It then also has a SlaveLock to serialize things (in case two builders are allowed to run concurrently on the same slave; each of the Git steps that touch the mirror's directory should run independently).